src/nti/schema/jsonschema.py

```diff
@@ -219,38 +222,66 @@ def make_schema(self):
-                'readonly': readonly,
-                'min_length': getattr(v, 'min_length', None),
-                'max_length': getattr(v, 'max_length', None)
-            }
-            ui_type = v.queryTaggedValue(TAG_UI_TYPE)
-            if not ui_type:
-                ui_type, ui_base_type = self.get_ui_types_from_field(v)
-            else:
-                _, ui_base_type = self.get_ui_types_from_field(v)
-
-            item_schema['type'] = ui_type
-            item_schema['base_type'] = ui_base_type
-
-            if sch_interfaces.IChoice.providedBy(v):
-                choices, base_type = self.get_data_from_choice_field(v, ui_base_type)
-                item_schema['choices'] = choices
-                item_schema['base_type'] = base_type
+            item_schema = self._make_field_schema(v)
 
             self.post_process_field(k, v, item_schema)
             ext_schema[k] = item_schema
 
         return ext_schema
+
+    def _translate(self, text):
+        return translate(text, context=self.context)
+
+    def _make_field_schema(self, field, name=None):
+        if field is None:
+            return {} # No constraints. Typical for a value_type or key_type
+
+        required = getattr(field, 'required', None)
+        required = field.queryTaggedValue(TAG_REQUIRED_IN_UI) or required
+        readonly_override = self.readonly_override
+        if readonly_override is not None:
+            readonly = readonly_override
+        else:
+            readonly = getattr(field, 'readonly', False)
+            readonly = field.queryTaggedValue(TAG_READONLY_IN_UI) or readonly
+
+        ui_base_type = None
+        item_schema = {
+            'name': name or field.__name__,
+            'required': required,
+            'readonly': readonly,
+            'min_length': getattr(field, 'min_length', None),
+            'max_length': getattr(field, 'max_length', None)
+        }
+
+        # Now things that we translate and presumably display to the user.
+        # Note that `field` is not necessarily an IField, it could be a simple
+        # Attribute
+        for attr in ('title', 'description'):
+            val = getattr(field, attr, None)
+            if val is not None:
+                item_schema[attr] = self._translate(val)
+
+
+        ui_type = field.queryTaggedValue(TAG_UI_TYPE)
+        if not ui_type:
+            ui_type, ui_base_type = self.get_ui_types_from_field(field)
+        else:
+            _, ui_base_type = self.get_ui_types_from_field(field)
+
+        item_schema['type'] = ui_type
+        item_schema['base_type'] = ui_base_type
+
+        if sch_interfaces.IChoice.providedBy(field):
+            choices, base_type = self.get_data_from_choice_field(field, ui_base_type)
+            item_schema['choices'] = choices
+            item_schema['base_type'] = base_type
+
+        if sch_interfaces.ICollection.providedBy(field):
+            item_schema['value_type'] = self._make_field_schema(field.value_type)
```

Comment:
Arggh, looking at this closer now. I think we'll want to build the full schema out for the `field.value_type`. I think right here we are externalizing the `Object` when perhaps we want to store `JsonSchemafier(field.value_type).make_schema()`.

Reply:
You can't do that. `JsonSchemafier` takes a schema, not a field. `value_type` and `key_type` are fields.

What you could do is check for fields that are `IObject` and recurse into *their* `.schema` attribute, which should be a schema.

Reply:
Yea, I believe you're correct. Sorry, when reviewing the PR, I thought it did what we needed.